### PR TITLE
New FPGA code sample "ac_fixed"

### PR DESCRIPTION
**Pull request by wangdi4**
# Adding a New Sample "ac_fixed"

## Description

This FPGA tutorial shows how to use the `ac_fixed` data type with some simple examples.

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): Tanner (@tyoungsc)

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by tyoungsc** [comment]
@wangdi4 I noticed you did a merge of master into your feature branch. In the future, try to use `git rebase` rather than `git merge`.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+#endif
+
+  try {
+    // Create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    // I. Constructing `ac_fixed` Numbers
+    // Rounding mode: use `AC_RND` to round towards plus infinity
+    // Overflow mode: use `AC_SAT` to saturate to max and min
+    ac_fixed<20, 10, true, AC_RND, AC_SAT> a;
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:126]: What are all these `AC_RND` and `AC_SAT`? Where do they come from? Some macros defined somewhere?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+void TestCalculateWithFloat(queue &q, const float &x, float &ret) {
+  buffer<float, 1> inp_buffer(&x, 1);
+  buffer<float, 1> ret_buffer(&ret, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp_buffer, h, read_only};
+    accessor res{ret_buffer, h, write_only, no_init};
+
+    h.single_task<CalculateWithFloat>([=] {
+      float sin_ret = sinf(x[0]);
```
Comment (pmpeter1) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:72]: For better readability would using sin_x (instead of sin_ret) and cos_x (instead of cos_ret) work better?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp (hunk 1 above)
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:126]: The problem is to understand where they come from.
Is `ac_fixed` defined in `::`? Same for `AC_RND` and `AC_SAT` or are they macros (argh!)? It does not fit well in a good looking C++ program.

**Review by tomlenth — APPROVED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp (hunk 1 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: `ac_fixed` is defined in `::`.
Looks like`AC_RND` and `AC_SAT` are enums in `::` (from ac_int.hpp around line 1234):

```cpp
enum ac_q_mode {
  AC_TRN,
  AC_RND,
  AC_TRN_ZERO,
  AC_RND_ZERO,
  AC_RND_INF,
  AC_RND_MIN_INF,
  AC_RND_CONV,
  AC_RND_CONV_ODD
};
```

A decent change may be moving from an `enum` to an `enum class`, but these files are from Mentor Graphics, so we many need to take up any changes with them. For now, it may be worthwhile adding to the comment where these values are coming from.

Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: Yes, and having `ac::trn` instead of `AC_TRN`, but this is another battle to have. :-)

Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:125]: agreed :|

**Review by pmpeter1 — DISMISSED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  W = 10, I = 3, S = true
+```
+
+In this tutorial, the `ac_fixed` numbers are smaller in size than floating point numbers, which results in a reduction of the FPGA resources at the expense of accuracy. To see the trade-offs in accuracy, compare the numeric results of the operations. The area utilization differences will be discussed in the section on *Examining the Reports*.
+
+When you use the `ac_fixed` library, keep the following points in mind:
+
+- Input Bit Width and Input Value Range Limits
+
+    The fixed-point math functions have bit width and input value range requirements. All bit width and input value range requirements are documented at the top of the `ac_fixed_math.hpp` files.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: Are they also documented in the documentation linked above? Where is a user going to find this file?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+- Return Types
+
+    For fixed-point functions, each function has a default return type. Assigning the result to a non-default return type triggers a type conversion and can cause an increase in logic use or a loss of accuracy in your results.  All return types are documented at the top of the `ac_fixed_math.hpp` file. For example, for `sin_fixed` and `cos_fixed`, the input type is `ac_fixed<W, 3, true>`, and the output type is `ac_fixed<W-1, 2, true>`. You can also use the `auto` type to let the compiler use the default return type.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:93]: Same remark as above

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+- Accuracy
+   - Floating point vs Fixed point
+
+     The host program for this tutorial shows the accuracy differences between the correct result and the result provided by the math library functions. The floating point version generates a more accurate result than the fixed point version.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:98]: Please precise that this is the case here because of the fixed-point datatype chosen for this example. One could have chosen a different ac_fixed datatype that would have resulted in a more accurate than floating point result

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+   ```bash
+   ./ac_fixed.fpga             # Linux
+   ```
+
+### Example of Output
+
+```txt
+Constructed from float:         3.6416015625
+Constructed from ac_fixed:      3.6416015625
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:266]: What are these "Constructed from" refer to? is this the result? If so, as the output is identical it is hard to understand below why there is a "MAX DIFF".
You can increase the number of digits to be displayed in order to show the difference between the two numbers.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+MAX DIFF for ac_fixed<10, 3, true>:  0.0078125
+MAX DIFF for float:                  9.53674e-07
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:269]: You could add a message saying: "computing on <x> values between <min> and <max>" followed by a message (not in full caps as none of the other messages are in full caps) saying "maximum difference observed between the golden value and the computed value" 
Also what is the methodology for knowing what is the golden value?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+result(float):       1
+
+result(fixed point): 1
+result(float):       1
+
+result(fixed point): 1
+result(float):       1
+
+result(fixed point): 0.992188
+result(float):       1
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:284]: For all these results, I don't know from this what they refer to, please add more print messages explaining what is beeing computed

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+PASSED: all kernel results are correct.
+```
+
+### Discussion of Results
+
+You can obtain a smaller hardware footprint for your kernel by ensuring that the `ac_fixed` numbers are constructed from `float` or `double` numbers outside the kernel. Additionally, you can trade-off mathematical operation accuracy for a more resource efficient design by using the `ac_fixed` math library functions.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:291]: Please add a comment saying that this is the case because the conversion will occur on the host rather than on the compute kernel.
And I would rephrase the second sentence as it suggests that the ac_fixed is less acccurate and more resource efficient no matter what which is not the case. I can create a fixed point format so large it covers the entire range of the floating-point format. In that case the resource utilisation is going to be way worse for exactly the same accuracy.
What is hard when using fixed-point numbers is knowing if your application could benefit from a fixed-point format, which is usually not the case.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/ac_fixed.vcxproj [hunk 10]
```diff
@@ -0,0 +1,164 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/ac_fixed.vcxproj:1]: Have you tested these MSVS files?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+# A DPC++ ahead-of-time (AoT) compile processes the device code in two stages.
+# 1. The "compile" stage compiles the device code to an intermediate representation (SPIR-V).
+# 2. The "link" stage invokes the compiler's FPGA backend before linking.
+#    For this reason, FPGA backend flags must be passed as link flags in CMake.
+set(EMULATOR_COMPILE_FLAGS "${WIN_FLAG} -fintelfpga ${AC_TYPES_FLAG} -DFPGA_EMULATOR -Wall")
+set(EMULATOR_LINK_FLAGS "-fintelfpga ${AC_TYPES_FLAG}")
+set(HARDWARE_COMPILE_FLAGS "${WIN_FLAG} -fintelfpga ${AC_TYPES_FLAG} -Wall")
+set(HARDWARE_LINK_FLAGS "-fintelfpga ${AC_TYPES_FLAG} -Xshardware -Xsboard=${FPGA_BOARD} ${USER_HARDWARE_FLAGS}")
+# We do not need to supply the AC_TYPES_FLAG for the 'report' target's linking stage.
+set(REPORT_LINK_FLAGS "-fintelfpga -Xshardware -Xsboard=${FPGA_BOARD} ${USER_HARDWARE_FLAGS}")
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/CMakeLists.txt:38]: This looks different than the other CMakeLists.txt files - do we need to specify here that we don't need the linker flag when building the report? Can we just instead use the hardware link flags? Just to keep all the cmake files as close as possible to one another

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 12]
```diff
@@ -0,0 +1,200 @@
+// clang-format off
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/ac_types/ac_fixed.hpp>
+#include <sycl/ext/intel/ac_types/ac_fixed_math.hpp>
+// clang-format on
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:6]: Why do we need these clang-format on/off comments here? 
I don't see us using these in the ac_int tutorial for example

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+class ConstructFromACFixed;
+class CalculateWithFloat;
+class CalculateWithACFixed;
+
+// Not recommended Usage example:
+// Convert dynamic float value inside the kernel
+void TestConstructFromFloat(queue &q, const float &a,
+                            ac_fixed<20, 10, true, AC_RND, AC_SAT> &b) {
+  buffer<float, 1> inp_buffer(&a, 1);
+  buffer<ac_fixed<20, 10, true, AC_RND, AC_SAT>, 1> ret_buffer(&b, 1);
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:29]: You made two type aliases above (fixed_10_3_t and fixed_9_2_t), why don't you have a third one with this datatype? It could also be a good to explain in this unique location what AC_RND and AC_SAT parameters are doing

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+// This is a FPGA best practice that reduces name mangling in the reports.
+class ConstructFromFloat;
+class ConstructFromACFixed;
+class CalculateWithFloat;
+class CalculateWithACFixed;
+
+// Not recommended Usage example:
+// Convert dynamic float value inside the kernel
+void TestConstructFromFloat(queue &q, const float &a,
+                            ac_fixed<20, 10, true, AC_RND, AC_SAT> &b) {
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:27]: Here the variable that holds the result is called b, but is called ret in TestConstructFromACFixed

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+    h.single_task<ConstructFromACFixed>([=] {
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> t(in_acc[0]);
+      ac_fixed<20, 10, true, AC_RND, AC_SAT> some_offset(0.5f);
+      out_acc[0] = t + some_offset;
+    });
+  });
+}
+
+void TestCalculateWithFloat(queue &q, const float &x, float &ret) {
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:63]: Here the input is called x where it is called a in the two above functions

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+// functions' type deduction rule. In this case, following those rules:
+// I, W, S are input type template parameter (ac_fixed<W, I, S>)
+// rI, rW, rS are output type template parameter (ac_fixed<rW, rI, rS>)
+//* Function Name         Type Propagation Rule
+//* sqrt_fixed             rI = I, rW = W, rS = S
+//* sin_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+//* cos_fixed              For signed (S == true), rI == 2, rW =  W - I + 2;
+//*                        For unsigned (S == false), I == 1, rW =  W - I + 1
+// clang-format on
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:90]: Why do you need clang-format on/off here?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+void TestCalculateWithACFixed(queue &q, const fixed_10_3_t &x,
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:91]: Input called x vs a in above functions, please use the same convention for all these functions

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+ac_fixed<4,0,false> x; // .xxxx : range = [0, 15/16], quantum = 1/16
+ac_fixed<4,7,false> x; // xxxx000. : range = [0, 120], quantum = 8
+ac_fixed<4,-3,true> x; // (.xxxx)2^(-3) : range=[-1/16, 7/128], quantum = 1/128
+```
+
+When creating an `ac_fixed` value, the created value may be less precise than the source value or the conversion may trigger an overflow. `ac_fixed` provides enums to select quantization and overflow modes, which determine how the source value will be converted.
+- The default quantization mode is `AC_TRN`, which drops bits to the right of LSB when quantization occurs.
+- The default overflow mode is `AC_WRAP`, which drops bits to the left of the MSB when overflow occurs.
+
+To select specific quantization and overflow modes, refer to section *2.1 Quantization and Overflow* in [Algorithmic C Datatypes Software Version v3.7](https://cdrdv2.intel.com/v1/dl/getContent/728986) for more details.
```
[the remark this answers is not in the document]
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:46]: It is now. Are you concerning about the version update of the document in the future?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md (hunk 3 above)
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: The document is in oneAPI optimization guide:  https://www.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide/top/optimize-your-design/resource-use/data-types-and-operations/var-prec-fp-sup/declare-and-use-the-ac-data-types/declare-the-ac-fixed-data-type.html,

which also refers to the `ac_fixed_math.hpp` file. Should we just mention this page in the tutorial or directly refer to the header file?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md (hunk 7 above)
Comment (wangdi4) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:269]: The `MAX DIFF` here is actually quantum instead of difference between values. Maybe we should also display the difference between the result and the golden value.

The golden value can be hardcoded

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 19]
```diff
@@ -86,7 +86,7 @@ When you use the `ac_fixed` library, keep the following points in mind:
 
 - Input Bit Width and Input Value Range Limits
 
-    The fixed-point math functions have bit width and input value range requirements. All bit width and input value range requirements are documented at the top of the `ac_fixed_math.hpp` files.
+    The fixed-point math functions have bit width and input value range requirements. All bit width and input value range requirements are documented at the top of the `ac_fixed_math.hpp` files, which locate in `${ONEAPI_ROOT}/compiler/latest/linux/lib/oclfpga/include/sycl/ext/intel/ac_types` on Linux or `%ONEAPI_ROOT%\compiler\latest\windows\lib\oclfpga\include\sycl\ext\intel\ac_types` on Windows.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:89]: files -> file

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md [hunk 20]
```diff
@@ -259,33 +259,50 @@ Scroll down on the Summary page of the report and expand the section titled **Co
...
+difference(float):              5.9604645e-08
 
 PASSED: all kernel results are correct.
 ```
 
 ### Discussion of Results
 
-You can obtain a smaller hardware footprint for your kernel by ensuring that the `ac_fixed` numbers are constructed from `float` or `double` numbers outside the kernel. Additionally, you can trade-off mathematical operation accuracy for a more resource efficient design by using the `ac_fixed` math library functions.
\ No newline at end of file
+You can obtain a smaller hardware footprint for your kernel by ensuring that the `ac_fixed` numbers are constructed from `float` or `double` numbers outside the kernel. Additionally, you can trade-off mathematical operation accuracy for a more resource efficient design by using `ac_fixed` types of a smaller size and the corresponding `ac_fixed` math library functions.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/README.md:308]: "of a smaller size" is not quite correct as you can use a larger fixed-point format than float and have arithmetic operators that use fewer resources than a floating point operator.
I would rephrase as follows: 
 "Additionally, by using the `ac_fixed` types and math library functions, you can use an even smaller fixed point format to trade-off even more accuracy for a more resource efficient design if your application requirements allows for it."

File: DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp [hunk 21]
```diff
@@ -1,19 +1,24 @@
...
 #include "dpc_common.hpp"
 
 using namespace sycl;
 
+// Type aliases for ac_fixed types
 using fixed_10_3_t = ac_fixed<10, 3, true>;
 using fixed_9_2_t = ac_fixed<9, 2, true>;
 
+// Quantization mode `AC_RND`: round towards plus infinity
+// Overflow mode `AC_SAT`: saturate to max and min when overflow happens
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/ac_fixed/src/ac_fixed.cpp:19]: Can you say here where a user can find the list of all the modes available?

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

**Review by yuguen-intel — APPROVED** [review]
Looks good to me, approving to get the CI flow running

**Comment by yuguen-intel** [comment]
The CI flow is currently disabled - I did the validation manualy. 
It is only missing @tomlenth and @pmpeter1 approval before merging
